pkg/fqdn/cache.go

```diff
@@ -0,0 +1,193 @@
+// This needs a read-lock
+func (s ipEntries) getIPs(now time.Time) []net.IP {
+	ips := make([]net.IP, 0, len(s)) // worst case size
+	for ip, entry := range s {
+		if entry != nil && !entry.isExpired(now) {
+			ips = append(ips, net.ParseIP(ip))
+		}
+	}
+
+	return keepUniqueIPs(ips) // sorts IPs
+}
+
+// DNSCache manages DNS data that will expire after a certain TTL. Information
+// is tracked per-IP address, retaining the latest-expiring DNS data for each
+// address.
+// For most real-world DNS data, the entry per name remains small because newer
+// lookups replace older ones. Large TTLs may cause entries to grow if many
+// unique IPs are returned in separate lookups.
+// Redundant or expired entries are removed on insert.
+// Lookups check for expired entries.
+type DNSCache struct {
+	lock.RWMutex
+
+	// forward DNS lookups name -> IPEntries
+	// IPEntries maps IP -> entry that provides it. An entry may provide multiple IPs.
+	forward map[string]ipEntries
+}
+
+// NewDNSCache returns an initialized DNSCache
+func NewDNSCache() *DNSCache {
+	c := &DNSCache{
+		forward: make(map[string]ipEntries),
+	}
+
+	return c
+}
+
+// Update inserts a new entry into the cache.
+// After insertion cache entries for name are expired and redundant entries
+// evicted. This is O(number of new IPs) for eviction, and O(number of IPs for
+// name) for expiration.
+func (c *DNSCache) Update(now time.Time, name string, ips []net.IP, ttl int) {
+	entry := &cacheEntry{
+		Name:           name,
+		LookupTime:     now,
+		ExpirationTime: now.Add(time.Duration(ttl) * time.Second),
+		TTL:            ttl,
+		IPs:            ips,
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	entries, exists := c.forward[name]
+	if !exists {
+		entries = make(map[string]*cacheEntry)
+		c.forward[name] = entries
+	}
+	c.updateWithEntryIPs(entries, entry)
+	c.removeExpired(entries, now)
+}
+
+// Lookup returns a set of unique IPs that are currently unexpired for name, if
+// any exist. An empty list indicates no valid records exist. The IPs are
+// returned sorted.
+func (c *DNSCache) Lookup(name string) (ips []net.IP) {
+	return c.lookupByTime(time.Now(), name)
+}
+
+// lookupByTime takes a timestamp for expiration comparisions, and is only
+// intended for testing.
+func (c *DNSCache) lookupByTime(now time.Time, name string) (ips []net.IP) {
+	c.RLock()
+	defer c.RUnlock()
+
+	entries, found := c.forward[name]
+	if !found {
+		return nil
+	}
+
+	return entries.getIPs(now)
+}
+
+// updateWithEntry adds a mapping for every IP in entry. It will replace
+// existing IP->old entry mappings if the older entry expires sooner (or has
+// already expired).
+// This needs a write lock
+func (c *DNSCache) updateWithEntryIPs(entries ipEntries, entry *cacheEntry) {
+	for _, ip := range entry.IPs {
+		ipStr := ip.String()
+		old, exists := entries[ipStr]
+		switch {
+		case old == nil || !exists:
+			entries[ipStr] = entry
+
+		case entry.ExpirationTime.After(old.ExpirationTime):
```

Comment:
I have a feeling this line is the opposite of this functions `func (entry *cacheEntry) isExpired(now time.Time) bool {`

Also, aren't all cases doing the same action in the end?

Reply:
They are the same! I have other code that needs to do something different in each case, but that is for another PR. I've made it an if, and switch to using `IsExpiredBy`.